enterprise/cmd/frontend/internal/completions/streaming/limiter.go

```diff
@@ -74,93 +73,52 @@ func (r *rateLimiter) TryAcquire(ctx context.Context) (err error) {
-	// like the limit and the time by when they should retry.
-	if current >= limit {
-		return RateLimitExceededError{
-			Limit:      limit,
-			Used:       current,
-			RetryAfter: retryAfter,
-		}
-	}
-
-	// Open a new connection to redis so we can run a MULTI command.
-	pool, ok := r.rstore.Pool()
-	if !ok {
-		return errors.New("redis pool is not available but rate limit has been configured")
-	}
-	conn, err := pool.Dial()
-	if err != nil {
-		return errors.Wrap(err, "failed to dial redis")
-	}
-	defer func() {
-		err = errors.Append(err, conn.Close())
-	}()
+	rstore := r.rstore.WithContext(ctx)
 
-	// Now that we validated that we want to let the user pass, let's increment
-	// the rate limit counter for the user.
+	// Let's increment the rate limit counter for the user.
 	// Note that the rate limiter _may_ allow slightly more requests than the configured
 	// limit, reading the current value and incrementing the rate limit counter are
 	// currently not an atomic operation.
-	if _, err := conn.Do("MULTI"); err != nil {
-		return errors.Wrap(err, "failed to start redis transaction")
-	}
-	// Increment the counter for the current user. If no record exists, redis will
-	// initialize it with 1.
-	if _, err := conn.Do("INCR", key); err != nil {
+	// We also don't use a transaction in here, because there is no good way to
+	// read the TTL without a lua script. This approach could slightly overcount the
+	// usage if redis requests after the INCR fail, but it will always recover safely.
+
+	// Check the current usage and increment the counter for the current user. If
+	// no record exists, redis will initialize it with 1.
+	currentUsage, err := rstore.Incr(key).Int()
+	if err != nil {
 		return errors.Wrap(err, "failed to increase rate limit counter")
 	}
 
 	// Set expiry on the key. If the key didn't exist prior to the previous INCR,
 	// it will set the expiry of the key to one hour.
-	// If it did exist before, it should have an expiry set already, so the NX makes
-	// sure that we don't overwrite it.
-	if _, err := conn.Do("EXPIRE", key, int(time.Hour/time.Second), "NX"); err != nil {
-		return errors.Wrap(err, "failed to set expiry for rate limit counter")
-	}
-
-	// Submit the calls.
-	if _, err := conn.Do("EXEC"); err != nil {
-		return errors.Wrap(err, "failed to flush multi operation")
-	}
-
-	return nil
-}
-
-func (r *rateLimiter) ReadQuota(ctx context.Context, userID int32) (count, limit int, _ error) {
-	limit, _, err := r.get(ctx, userKey(userID))
+	// If it did exist before, it should have an expiry set already, so the TTL > 0
+	// makes sure that we don't overwrite it and restart the 1h bucket.
+	ttl, err := rstore.TTL(key)
 	if err != nil {
-		return 0, 0, err
+		return errors.Wrap(err, "failed to get TTL for rate limit counter")
 	}
-	configuredLimit, err := getConfiguredLimit(actor.WithActor(context.Background(), actor.FromUser(userID)), r.db)
-	if err != nil {
-		return 0, 0, err
+	if ttl <= 0 {
```

Comment:
as mentioned in the comment above, I think this should be
```suggestion
	if ttl < 0 {
```